modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/KeyEventFirerTest.java

```diff
@@ -0,0 +1,171 @@
+
+    private TextField textField;
+    private Button button;
+    private Pane root;
+    private Stage stage;
+    private Scene scene;
+
+    /**
+     * Test that keyEvent is delivered to focused control and nowhere else.
+     */
+    @Test
+    public void testFireViaScene() {
+        showAndFocus(button);
+        List<KeyEvent> buttonEvents = new ArrayList<>();
+        button.addEventHandler(KEY_PRESSED, buttonEvents::add);
+        List<KeyEvent> textFieldEvents = new ArrayList<>();
+        textField.addEventHandler(KEY_PRESSED, textFieldEvents::add);
+        KeyEventFirer firer = new KeyEventFirer(textField, scene);
+        firer.doKeyPress(A);
+        assertEquals("button must have received the key", 1, buttonEvents.size());
+        assertEquals("textField must not have received the key", 0, textFieldEvents.size());
+    }
+
+    /**
+     * Test that keyEvent is delivered to focused control and nowhere else.
+     * Here we test that the target is not required.
+     */
+    @Test
+    public void testFireViaSceneNullTarget() {
+        showAndFocus(button);
+        List<KeyEvent> buttonEvents = new ArrayList<>();
+        button.addEventHandler(KEY_PRESSED, buttonEvents::add);
+        List<KeyEvent> textFieldEvents = new ArrayList<>();
+        textField.addEventHandler(KEY_PRESSED, textFieldEvents::add);
+        KeyEventFirer firer = new KeyEventFirer(null, scene);
+        firer.doKeyPress(A);
+        assertEquals("button must have received the key", 1, buttonEvents.size());
+        assertEquals("textField must not have received the key", 0, textFieldEvents.size());
+    }
+
+    /**
+     * This simulates a false positive: even though not focused, the textField handlers
+     * are notified when firing directly. That's possible, but typically not what we want to test!
+     */
+    @Test
+    public void testFireTargetFalseGreen() {
+        showAndFocus(button);
+        List<KeyEvent> buttonEvents = new ArrayList<>();
+        button.addEventHandler(KEY_PRESSED, buttonEvents::add);
+        List<KeyEvent> textFieldEvents = new ArrayList<>();
+        textField.addEventHandler(KEY_PRESSED, textFieldEvents::add);
+        // firing on a node that is not focusOwner
+        KeyEventFirer incorrectFirer = new KeyEventFirer(textField);
+        incorrectFirer.doKeyPress(A);
+        int falseTextFieldNotification = textFieldEvents.size();
+        int falseButtonNotification = buttonEvents.size();
+        assertEquals("false green - textField must have received the key", 1, textFieldEvents.size());
+        assertEquals("false green - button must not have received the key", 0, buttonEvents.size());
+        textFieldEvents.clear();
+        buttonEvents.clear();
+        // firing on the scene makes a difference
+        KeyEventFirer correctFirer = new KeyEventFirer(null, scene);
+        correctFirer.doKeyPress(A);
+        assertNotEquals(falseTextFieldNotification, textFieldEvents.size());
```

Comment:
assertNotEquals does not compile for me.

Reply:
It looks like `assertNotEquals` was added in a later version of junit than the one we use, which is still junit 4.8.2. We should upgrade to a newer version at some point, but until then we need to avoid using anything from newer versions of junit.

Reply:
*I-must-not-push-without-commandline-build, *I-must-not-push-without-commandline-build, *I-must-...

sry, guys, for wasting your time - changed